### DISCUSSION TITLE
Keep window above all other windows

### DISCUSSION THREAD

**Discussion by ervinpopescu** [qtile/qtile#4049] [Q&A] [answered]
I'm trying to keep `plank` above all windows at all times (well, except when a window is fullscreen). As far as I can tell from [this](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#idm45154644220640), this is what I want.

How would one do this?

Right now, I'm doing this:
```py
@hook.subscribe.client_new
@hook.subscribe.client_focus
@hook.subscribe.client_managed
@hook.subscribe.client_killed
def bring_plank_to_front(_window):
    if _window.window.get_wm_class()[0] != "nwggrid-server":
        subprocess.Popen([os.path.join(qtile_path, "scripts", "bring_plank_to_front.py")])
```
where `bring_plank_to_front.py` is:
```py
#!/bin/python

from libqtile.command.client import InteractiveCommandClient

c = InteractiveCommandClient()

windows = c.windows()
for window in windows:
    if window["name"] == "plank":
        id = window["id"]
        c.window[id].bring_to_front()
```

Something wasn't working in the config, so that's why my lazy person moved it to a script.
Oh and one more thing, this 'hack' isn't cool with me or me with it because it doesn't follow my original 'requirement' that `plank` will not show up above fullscreen windows. And then there's also [this](https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#STACKINGORDER).

This is what happens when I don't use the script:
https://user-images.githubusercontent.com/84532402/206884851-410a23f3-ecc9-4e11-88ce-e91fcbe40c00.mp4

Relevant stuff:
```py
@hook.subscribe.client_new
@hook.subscribe.client_focus
@hook.subscribe.client_managed
@hook.subscribe.client_killed
@lazy.function
def bring_plank_to_front(client, qtile):
    if client.window.get_wm_class()[0] != "nwggrid-server":
        windows = qtile.windows()
        for window in windows:
            if window["name"] == "plank":
                id = window["id"]
                qtile.current_screen.window[id].bring_to_front()
```

I don't think `lazy.function` works here, or what I'm supposed to do to make this approach work.

Sorry for the wall of text and (1) video :)

```console
$ xprop -name plank
_NET_WM_STATE(ATOM) = _NET_WM_STATE_ABOVE
WM_CLASS(STRING) = "plank", "Plank"
WM_COMMAND(STRING) = { "plank" }
WM_CLIENT_LEADER(WINDOW): window id # 0x2400001
_NET_WM_PID(CARDINAL) = 1372
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "arch"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified size: 20 by 20
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING
WM_ICON_NAME(STRING) = "plank"
_NET_WM_ICON_NAME(UTF8_STRING) = "plank"
WM_NAME(STRING) = "plank"
_NET_WM_NAME(UTF8_STRING) = "plank"
```

Trying to set it with xprop:
```sh
xprop -name plank -f _NET_WM_STATE 32a -set _NET_WM_STATE '_NET_WM_STATE_STICKY, _NET_WM_STATE_SKIP_TASKBAR, _NET_WM_STATE_SKIP_PAGER, _NET_WM_STATE_ABOVE'
```
and
```console
$ xprop -name plank
_NET_WM_STATE(ATOM) = _NET_WM_STATE_STICKY, _NET_WM_STATE_SKIP_TASKBAR, _NET_WM_STATE_SKIP_PAGER, _NET_WM_STATE_ABOVE
WM_CLASS(STRING) = "plank", "Plank"
WM_COMMAND(STRING) = { "plank" }
WM_CLIENT_LEADER(WINDOW): window id # 0x2400001
_NET_WM_PID(CARDINAL) = 1372
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "arch"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified size: 20 by 20
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING
WM_ICON_NAME(STRING) = "plank"
_NET_WM_ICON_NAME(UTF8_STRING) = "plank"
WM_NAME(STRING) = "plank"
_NET_WM_NAME(UTF8_STRING) = "plank"
```
It appears I have to change the title :)

**Comment by ervinpopescu** [comment]
Wait, is this related to https://github.com/qtile/qtile/pull/3409? 😿

**Reply by elParaguayo** [reply]
Yes - this is the kind of thing that that PR should address. Ideally, if `plank` is set to be above windows, qtile should respect that.

I'm planning on prioritising that PR as it would solve a lot of issues for people.

**Reply by ervinpopescu** [reply]
One more question please.
How do I use the `qtile` object in hooks again?

**Reply by elParaguayo** [reply]
What do you mean? If you mean how can you access the qtile object, you just need to do `from libqtile import qtile`

**Reply by ervinpopescu** [reply]
Actually, I have one more. :)
Why is this happenning (no more _NET_WM_STATE) ?
```console
$ xprop -name plank
WM_CLASS(STRING) = "plank", "Plank"
WM_COMMAND(STRING) = { "plank" }
WM_CLIENT_LEADER(WINDOW): window id # 0x2800001
_NET_WM_PID(CARDINAL) = 3133
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "arch"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified size: 20 by 20
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING
WM_ICON_NAME(STRING) = "plank"
_NET_WM_ICON_NAME(UTF8_STRING) = "plank"
WM_NAME(STRING) = "plank"
_NET_WM_NAME(UTF8_STRING) = "plank"
```

**Reply by ervinpopescu** [reply]
> What do you mean? If you mean how can you access the qtile object, you just need to do `from libqtile import qtile`

Oh, so why didn't this work?
```py
@hook.subscribe.client_new
@hook.subscribe.client_focus
@hook.subscribe.client_managed
# @hook.subscribe.client_killed # oops, this one isn't right :)
@lazy.function
def bring_plank_to_front(client, qtile):
    if client.window.get_wm_class()[0] != "nwggrid-server":
        windows = qtile.windows()
        for window in windows:
            if window["name"] == "plank":
                id = window["id"]
                qtile.current_screen.window[id].bring_to_front()
```

I'll try without the `lazy.function` hook and the `qtile` parameter.

I get this in the log:
```log
2022-12-11 17:22:32,349 ERROR libqtile hook.py:fire():L406 Error in hook client_focus
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/libqtile/hook.py", line 404, in fire
    i(*args, **kwargs)
  File "/home/ervin/.config/qtile-x11/modules/hooks.py", line 58, in bring_plank_to_front
    qtile.current_screen.window[id].bring_to_front()
  File "/usr/lib/python3.10/site-packages/libqtile/command/base.py", line 283, in __getattr__
    raise AttributeError
AttributeError
```

**Reply by elParaguayo** [reply]
I'm not sure you can use all those decorators together. `lazy.function` will pass `qtile` as the first argument which means `client` will have the value of `qtile`.

Anyway, as for your loop, you can make it much simpler. Something like this should do what you want:
``` python
for window in qtile.windows_map.values():
    if window.name == "plank":
        window.bring_to_front()
```

**Reply by ervinpopescu** [reply]
Yep. That works. I'm not gonna use the comment I linked, it's not working properly, I'll use this instead. Thanks!

**Comment by ervinpopescu** [comment, marked as answer]
This solves my problem, for now:
https://github.com/qtile/qtile/issues/1145#issuecomment-392771986

Still waiting for #3409 tho, great stuff :)

**Reply by ervinpopescu** [reply]
How exactly am I to use #3409 now regarding this question, @elParaguayo?
Tried:
```python
from libqtile.command.client import InteractiveCommandClient
from libqtile.core.manager import Qtile
c: Qtile = InteractiveCommandClient()

for window in c.windows():
    if window["name"] == "plank":
        c.window[window["id"]].keep_above()
```
and the plank window does get `_NET_WM_STATE_ABOVE`, but nothing happens with the window. If I use https://github.com/ervinpopescu/dots/blob/lenovo/.config/qtile/modules/hooks.py#L52-L58, that works, but AFAICT from https://docs.qtile.org/en/latest/manual/config/lazy.html#window-functions, `bring_to_front` ignores the `kept_above`.

**Reply by elParaguayo** [reply]
What's the question?

**Reply by ervinpopescu** [reply]
How do I use the new functions, basically.

**Reply by elParaguayo** [reply]
Sorry for the delay. All the commands are available via `lazy.window...` and so can be bound to keys etc.

The commands are all shown here: https://docs.qtile.org/en/latest/manual/commands/api/windows.html

EDIT: Sorry - I've misread your question. Are you saying that, despite setting `keep_above`, other windows show above `plank`?

**Reply by ervinpopescu** [reply]
Yes, unless I do the [hacky hack](https://github.com/ervinpopescu/dots/blob/lenovo/.config/qtile/modules/hooks.py#L52-L58).

**Reply by elParaguayo** [reply]
OK - can you start an issue for this and I'll look later.
